rust/chains/hyperlane-cosmos/src/providers/mod.rs

```diff
@@ -172,14 +204,22 @@ impl HyperlaneProvider for CosmosProvider {
             .map_err(|e| ChainCommunicationError::from_other_str("could not convert from proto"))?;
         let contract = H256::try_from(CosmosAccountId::new(&msg.contract))?;
 
-        // TODO choose correct signer (should it be the one paid for the transaction)
-        let signer = tx
+        let (sender, nonce) = tx
             .auth_info
-            .signer_infos
-            .get(0)
-            .expect("there should be at least one signer");
-        let sender = self.sender(signer)?;
-        let nonce = signer.sequence;
+            .fee
+            .payer
+            .as_ref()
+            .map(|payer| self.search_payer_in_signer_infos(&tx.auth_info.signer_infos, payer))
+            .map_or_else(
+                || {
+                    let signer_info = tx.auth_info.signer_infos.get(0).ok_or_else(|| {
+                        ChainCommunicationError::from_other_str("no signer info in default signer")
```

Comment:
something we'll have to think about is general error handling in the scraper and how fragile it is. Atm the scraper prefers to not insert any data to the db if there are errors getting things like blocks / txs relating to that piece data. Some context here https://github.com/hyperlane-xyz/issues/issues/1369

It doesn't have to be in this PR, but I think we may want to rethink this. E.g. if there's some special transaction on Cosmos that for some reason doesn't have any fee payer or signers (maybe some "system transaction"), atm the scraper will propagate this error such that the piece of data that really matters, e.g. a message that was indexed in that transaction, ultimately doesn't get inserted into the db.

I wonder instead if some things should just be optional in the DB - e.g. if the sender wasn't able to be found by our logic, that's fine and we just log it and write null to the DB, but we still write the Hyperlane message that occurred in that transaction to the DB

Reply:
I agree. We need to revisit it in more general settings thought.